model/riscv_fdext_regs.sail

```diff
@@ -6,64 +6,100 @@
 /*  SPDX-License-Identifier: BSD-2-Clause                                                */
 /*=======================================================================================*/
 
-/* **************************************************************** */
-/* Floating point register file and accessors for F, D extensions   */
-/* Floating point CSR and accessors                                 */
-/* **************************************************************** */
+/* ***************************************************************** */
+/* Floating point register file and accessors for F, D, Q extensions */
+/* Floating point CSR and accessors                                  */
+/* ***************************************************************** */
 
 /* Original version written by Rishiyur S. Nikhil, Sept-Oct 2019    */
 
 /* **************************************************************** */
 /* NaN boxing/unboxing.                                             */
-/* When 16-bit floats (half-precision) are stored in 32/64-bit regs  */
-/* they must be 'NaN boxed'.                                         */
-/* When 16-bit floats (half-precision) are read from 32/64-bit regs  */
-/* they must be 'NaN unboxed'.                                       */
-/* When 32-bit floats (single-precision) are stored in 64-bit regs  */
+/* When 16-bit floats (half-precision) are stored in 32/64/128-bit  */
+/* regs they must be 'NaN boxed'.                                   */
+/* When 16-bit floats (half-precision) are read from 32/64/128-bit  */
+/* regs they must be 'NaN unboxed'.                                 */
+/* When 32-bit floats (single-precision) are stored in 64/128-bit   */
+/* regs they must be 'NaN boxed'.                                   */
+/* When 32-bit floats (single-precision) are read from 64/128-bit   */
+/* regs they must be 'NaN unboxed'.                                 */
+/* When 64-bit floats (double-precision) are stored in 128-bit regs */
 /* they must be 'NaN boxed' (upper 32b all ones).                   */
-/* When 32-bit floats (single-precision) are read from 64-bit regs  */
+/* When 64-bit floats (double-precision) are read from 128-bit regs */
 /* they must be 'NaN unboxed'.                                      */
 
 function canonical_NaN_H() -> bits(16) = 0x_7e00
 function canonical_NaN_S() -> bits(32) = 0x_7fc0_0000
 function canonical_NaN_D() -> bits(64) = 0x_7ff8_0000_0000_0000
+function canonical_NaN_Q() -> bits(128) = 0x_7fff_0000_0000_0000_0000_0000_0000_0000
 
 val      nan_box_H : bits(16) -> flenbits
 function nan_box_H   val_16b =
-  if (sizeof(flen) == 32)
-  then 0x_FFFF @ val_16b
-  else 0x_FFFF_FFFF_FFFF @ val_16b
+  match (sizeof(flen)) {
```

Comment:
I know there is work being done on quad tests for riscv-arch-test which will want to use the sail model to generate signatures. If those are completed before we get all of the floating point updates into the Sail standard library and make the code more generic, then we may want to revisit merging this sooner but otherwise, I agree.

Reply:
@jordancarlin Do you know if anyone is planning to make and tape-out a real chip that implements the Q extension?

Reply:
@martinberger The OpenHW Core-V-Wally processor (https://github.com/openhwgroup/cvw) implements the Q extension in some of its configurations. There is an upcoming tape-out planned for it very soon. I'm double checking which configuration is planned for that, but it sounds like it might include Q.